apps/emqx_authz/src/emqx_authz_api_schema.erl

```
@@ -16,490 +16,185 @@
+    lists:all(
+      fun erlang:is_binary/1,
+      maps:values(Body)).
+
+%%------------------------------------------------------------------------------
+%% MonogDB type funcs
+
+authz_mongo_common_fields() ->
+    authz_common_fields(mongodb) ++
+    [ {collection, fun collection/1}
+    , {selector, fun selector/1}
     ].
+
+collection(type) -> binary();
+collection(_) -> undefined.
+
+selector(type) -> map();
+selector(_) -> undefined.
+
+%%------------------------------------------------------------------------------
+%% Redis type funcs
+
+authz_redis_common_fields() ->
+    authz_common_fields(redis) ++
+    [ {cmd, #{ type => string()
+             , example => <<"HGETALL mqtt_authz">>}}].
+
+%%------------------------------------------------------------------------------
+%% Authz api type funcs
+
+authz_common_fields(Type) when is_atom(Type)->
+    [ {enable, fun enable/1}
+    , {type, #{ type => enum([Type])
+              , default => Type
+              , converter => fun to_bin/1}}
+    ].
+
+enable(type) -> boolean();
+enable(default) -> true;
+enable(desc) -> "Set to <code>false</code> to disable this auth provider";
+enable(_) -> undefined.
+
+%%------------------------------------------------------------------------------
+%% Internal funcs
+
+authz_sources_types(Type) ->
+    case Type of
+        simple -> [mongodb, redis];
+        detailed -> [ mongo_single
+                    , mongo_rs
+                    , mongo_sharded
+                    , redis_single
+                    , redis_sentinel
+                    , redis_cluster]
+    end
+        ++
+        [ http
+        , 'built-in-database'
+        , mysql
+        , postgresql
+        , file].
+
+to_list(A) when is_atom(A) ->
+    atom_to_list(A);
+to_list(B) when is_binary(B) ->
+    binary_to_list(B).
+
+to_bin(List) when is_list(List) ->
+    case io_lib:printable_list(List) of
+        true -> unicode:characters_to_binary(List);
+        false -> List
```

Let this crash? 
maybe we never enter this case?
we should always ensure the result is binary.